[feature] new semnidaq driver: control a SEM via the NI-DAQ API

pieleric opened the pull request:
Depends on the "nidaqmx-python" Python wrapper for using the NI-DAQmx API.
For now tested only on a NI PCIe-6361... and also known to not work
yet on older devices like the PCIe-6251.

It supports all features needed for the SECOM and SPARC, with the
exception of the counter detector. Tested succesfully on a real SEM.

It is possible to test it on a computer without an actual card by using
the NI-DAQmx with a simulated DAQ card. Instructions are in the top
comment of semnidaq.py .

Also introduces two microscope files: one to use a real DAQ board, only,
and one to simulate a SPARC using a simulated NI DAQ board.A

pieleric committed:
[feature] new semnidaq driver: control a SEM via the NI-DAQ API

Depends on the "nidaqmx-python" Python wrapper for using the NI-DAQmx API.
For now tested only on a NI PCIe-6361... and also known to not work
yet on older devices like the PCIe-6251.

It supports all features needed for the SECOM and SPARC, with the
exception of the counter detector. Tested succesfully on a real SEM.

It is possible to test it on a computer without an actual card by using
the NI-DAQmx with a simulated DAQ card. Instructions are in the top
comment of semnidaq.py .

Also introduces two microscope files: one to use a real DAQ board, only,
and one to simulate a SPARC using a simulated NI DAQ board.A

FCanberk reviewed (commented):
i am doing the new linter's job for this PR: 

please don't forget to add the empty new lines at the end of files  :grin:

tepals reviewed (changes requested):
I tried to look into everything as much as possible, but I'm sure I missed something. Let me know if there is a specific part or function that you want me to look into in more detail.

pieleric commented on src/odemis/driver/semnidaq.py:
```diff
@@ -0,0 +1,2799 @@
+        :param children: (internal role -> kwargs) parameters setting for the children.
+            Known children are "scanner", "detector0", "detector1"...
+            They will be provided back in the .children roattribute
+        :param device: name of the NI DAQ device (ex: "Dev1"). See "nilsdev".
```
It's a command. I've updated the comment.

pieleric commented on src/odemis/driver/semnidaq.py:
```diff
@@ -0,0 +1,2799 @@
+        system = nidaqmx.system.System.local()
+
+        try:
+            self._nidev = system.devices[device]  # This accepts anything. It will fail on first call
+            self._hwVersion = f"NI {self._nidev.product_type} s/n: {self._nidev.serial_num}"  # Will fail here
```
Indeed... not very clear. It fails *if it's incorrect* only on line 142. Updated the comment.

pieleric commented on src/odemis/driver/semnidaq.py:
```diff
@@ -0,0 +1,2799 @@
+        roi_limits_raw = []  # min/max for X/Y in raw value (int16)
+        for i, lim in enumerate(self._limits):
+            center = (lim[0] + lim[1]) / 2
+            width = lim[1] - lim[0]
```
Not necessarily. But it works in both directions. I think the word is badly picked though, as it implies a positive value. => Changed to `volt_diff`.

pieleric committed:
[feature] add util.round_up_to_multiple() function

Allows to round a value to multiple of any given number, while managing
floating point error.

pieleric commented on src/odemis/driver/semnidaq.py:
```diff
@@ -0,0 +1,2799 @@
+            # Note that it may still give warnings
+            self._multi_detector_min_period = 1 / self._nidev.ai_max_multi_chan_rate
+        elif not isinstance(multi_detector_min_period, (float, int)):
+            raise ValueError(f"multi_detector_min_period must be float, but is {type(multi_detector_min_period)}")
```
I've adjusted it to just say "a number".

pieleric commented on install/linux/usr/share/odemis/sim/sparc2-nidaq-sim.odm.yaml:
```diff
@@ -0,0 +1,424 @@
+# Light (lamp with known spectrum)
+"Calibration Light": {
+    class: light.Light,
+    role: "brightlight",
```
With and without quotes are equivalent, as in YAML, strings are automatically detected. It's the same for the classes or the component names. However strings with spaces must be quoted. So we have the habit to always quote the component names. For classes and role, they are not supposed to contain spaces, so it's safe to never quote them. I've cleaned up this microscope file to drop all the quotes from the "role".
